libsolidity/formal/BMC.cpp

```diff
@@ -290,109 +291,206 @@ bool BMC::visit(Conditional const& _op)
-// executed once before the condition, might reassign variables.
-// Variables touched by the loop are merged with Branch 2.
+// Unrolls while or do-while loop
 bool BMC::visit(WhileStatement const& _node)
 {
-	auto indicesBeforeLoop = copyVariableIndices();
+	auto indicesBefore = copyVariableIndices();
+
+	m_context.pushSolver();
+	// variables touched by loop might change their value
+	// assume their values are not constant
 	m_context.resetVariables(touchedVariables(_node));
-	decltype(indicesBeforeLoop) indicesAfterLoop;
+
+	_node.condition().accept(*this);
+
+	// do not apply condition target on expressions
+	// that value might be changed by other loops
+	if (isRootFunction() && !isInsideLoop())
+		addVerificationTarget(
+			VerificationTargetType::ConstantCondition,
+			expr(_node.condition()),
+			&_node.condition()
+		);
+	m_context.popSolver();
+
+	resetVariableIndices(indicesBefore);
+
+	unsigned int bmcLoopIterations = m_settings.bmcLoopIterations.value_or(1);
+	smtutil::Expression broke(false);
+	smtutil::Expression loopCondition(true);
+	unsigned int start = 0;
 	if (_node.isDoWhile())
 	{
-		indicesAfterLoop = visitBranch(&_node.body()).first;
-		// TODO the assertions generated in the body should still be active in the condition
+		m_loopCheckpoints.emplace();
+
+		auto indicesBefore = copyVariableIndices();
+		_node.body().accept(*this);
 		_node.condition().accept(*this);
-		if (isRootFunction())
-			addVerificationTarget(
-				VerificationTargetType::ConstantCondition,
-				expr(_node.condition()),
-				&_node.condition()
-			);
+
+		auto [continues, brokeInCurrentIteration] =
+			mergeVariablesFromLoopCheckpoints();
+
+		acceptLoopConditionOnContinueStatement(_node, continues);
+
+		broke = broke || brokeInCurrentIteration;
+		m_loopCheckpoints.pop();
+		start = 1;
 	}
 	else
-	{
 		_node.condition().accept(*this);
-		if (isRootFunction())
-			addVerificationTarget(
-				VerificationTargetType::ConstantCondition,
-				expr(_node.condition()),
-				&_node.condition()
-			);
 
-		indicesAfterLoop = visitBranch(&_node.body(), expr(_node.condition())).first;
-	}
+	for (unsigned int i = start; i < bmcLoopIterations; ++i)
+	{
+		m_loopCheckpoints.emplace();
+
+		auto indicesBefore = copyVariableIndices();
+		loopCondition = expr(_node.condition());
 
-	// We reset the execution to before the loop
-	// and visit the condition in case it's not a do-while.
-	// A do-while's body might have non-precise information
-	// in its first run about variables that are touched.
-	resetVariableIndices(indicesBeforeLoop);
-	if (!_node.isDoWhile())
+		pushPathCondition(loopCondition);
+		_node.body().accept(*this);
 		_node.condition().accept(*this);
+		popPathCondition();
+
+		auto [continues, brokeInCurrentIteration] =
+			mergeVariablesFromLoopCheckpoints();
 
-	mergeVariables(expr(_node.condition()), indicesAfterLoop, copyVariableIndices());
+		acceptLoopConditionOnContinueStatement(_node, continues);
```

Comment:
I don't quite understand why this is done here and needs a merge inside the function.
Why can't we simply do:
- do not visit the condition in the `do-while` if/else above
- always visit the condition as the first thing inside this for loop, similar to how it's implemented below for `ForLoop`